test/qdmi/test_driver.cpp

```diff
@@ -0,0 +1,379 @@
+      << "Devices must implement `QDMI_device_query_operation_property`.";
+}
+
+TEST_P(DriverTest, QueryDeviceVersion) {
+  size_t size = 0;
+  ASSERT_EQ(QDMI_device_query_device_property(
+                device, QDMI_DEVICE_PROPERTY_VERSION, 0, nullptr, &size),
+            QDMI_SUCCESS)
+      << "Devices must provide a version.";
+  std::string value(size - 1, '\0');
+  ASSERT_EQ(QDMI_device_query_device_property(device,
+                                              QDMI_DEVICE_PROPERTY_VERSION,
+                                              size, value.data(), nullptr),
+            QDMI_SUCCESS)
+      << "Devices must provide a version.";
+  EXPECT_FALSE(value.empty()) << "Devices must provide a version.";
+}
+
+TEST_P(DriverTest, QueryDeviceLibraryVersion) {
+  size_t size = 0;
+  ASSERT_EQ(QDMI_device_query_device_property(
+                device, QDMI_DEVICE_PROPERTY_LIBRARYVERSION, 0, nullptr, &size),
+            QDMI_SUCCESS)
+      << "Devices must provide a library version.";
+  std::string value(size - 1, '\0');
+  ASSERT_EQ(QDMI_device_query_device_property(
+                device, QDMI_DEVICE_PROPERTY_LIBRARYVERSION, size, value.data(),
+                nullptr),
+            QDMI_SUCCESS)
+      << "Devices must provide a library version.";
+  ASSERT_FALSE(value.empty()) << "Devices must provide a library version.";
+}
+
+TEST_P(DriverTest, QueryNumQubits) {
+  size_t numQubits = 0;
+  ASSERT_EQ(
+      QDMI_device_query_device_property(device, QDMI_DEVICE_PROPERTY_QUBITSNUM,
+                                        sizeof(size_t), &numQubits, nullptr),
+      QDMI_SUCCESS)
+      << "Devices must provide the number of qubits.";
+  EXPECT_GT(numQubits, 0) << "Number of qubits must be greater than 0.";
+}
+
+TEST_P(DriverTest, QueryDeviceProperties) {
+  EXPECT_EQ(QDMI_device_query_device_property(device, QDMI_DEVICE_PROPERTY_MAX,
+                                              0, nullptr, nullptr),
+            QDMI_ERROR_INVALIDARGUMENT)
+      << "The MAX property is not a valid value for any device.";
+}
+
+TEST_P(DriverTest, QuerySites) {
+  size_t size = 0;
+  ASSERT_EQ(QDMI_device_query_device_property(
+                device, QDMI_DEVICE_PROPERTY_SITES, 0, nullptr, &size),
+            QDMI_SUCCESS)
+      << "Devices must provide a list of sites.";
+  std::vector<QDMI_Site> sites(size / sizeof(QDMI_Site));
+  ASSERT_EQ(QDMI_device_query_device_property(
+                device, QDMI_DEVICE_PROPERTY_SITES, size,
+                static_cast<void*>(sites.data()), nullptr),
+            QDMI_SUCCESS)
+      << "Failed to get sites.";
+  std::unordered_set<size_t> ids;
+  for (auto* site : sites) {
+    uint64_t id = 0;
+    EXPECT_EQ(QDMI_device_query_site_property(device, site,
+                                              QDMI_SITE_PROPERTY_INDEX,
+                                              sizeof(uint64_t), &id, nullptr),
+              QDMI_SUCCESS)
+        << "Devices must provide a site id";
```

Comment:
IDs as a term should be replaced by "index" here as this was renamed in QDMI as well.